File: eulerr.R
#eulerr command line script

#pass in a data table with columns:
# - intersection name
# - intersection count
# - intersection color

# alternative:
# pass in a CSV with named logical columns representing the presence or absence of an item in the set.

# how to handle toggling labels/quantities?

args <- commandArgs(TRUE)

setsfile <- args[1] 
shape <- args[2] 
alpha <- as.numeric(args[3]) 
width <- as.numeric(args[4])
height <- as.numeric(args[5]) 
fsz_counts <- as.numeric(args[6])
fsz_labels <- as.numeric(args[7])
rng_seed <- as.numeric(args[8])
pdf_file <- args[9]

sets_df <-read.csv(setsfile, header = T)

combos <- setNames(unlist(sets_df$counts), unlist(sets_df$names))


control <- list(extraopt=T,
                extraopt_threshold = 0,
                extraopt_control = list(seed = rng_seed, max.time=30,  
                                        verbose=T, smooth=F, temperature=1000))
# list(seed = rng_seed, smooth=F, 
#      maxit=1e100, temperature=10000,verbose=T,
#      nb.stop.improvement=1e3)
print(control)

fit1<-eulerr::euler(combos, shape=shape, control=control)

print(fit1)

fills <- list()
if ("cols" %in% colnames(sets_df) ) {
  fills$fill = unlist(sets_df$cols)
}
# print(fills)

fills$alpha = alpha

quantities <- FALSE
if (fsz_counts>0) {
  quantities <- list(fontsize=fsz_counts)
}

labels <- FALSE
if (fsz_labels>0) {
  labels <- list(fontsize=fsz_labels)
}

pdf_width <- NULL
if (width>0) {
  pdf_width <- width
}
pdf_height <- NULL
if (height>0) {
  pdf_height <- height
}


pdf(pdf_file, width = pdf_width, height = pdf_height)
plot(fit1, 
     quantities = quantities, 
     labels = labels,
     fills = fills)
dev.off()

File: eulerr.m
function [excl_int,int_names]=eulerr(sets, setnames, pdf_file, options)
arguments
    sets
    setnames
    pdf_file
    options.do_int = true
    options.shape = "circle"
    options.cols = []
    options.alpha = 1
    options.fsz_counts = 8
    options.fsz_labels = 8
    options.width = 0
    options.height = 0
    options.rng_seed = 42
    options.verbose=false
end
%set fsz to 0 to not display the counts or labels

if ~options.do_int 
    counts=sets;
    int_names=setnames;
else
    setnames=string(setnames);
    [excl_int,~,int_names]=get_exclusive_intersections(sets, setnames);
    int_names=string(int_names);
    int_names=strrep(int_names,'_','&');
    counts=cellfun(@length,excl_int);
end

T=table();
T.names=int_names(:);
T.counts=counts(:);

if ~isempty(options.cols)
    hexcols=rgb2hex(options.cols);
    hexcols=string(hexcols);
    T.cols=hexcols(:);
end

% T(counts==0,:)=[];

setsfile="tmp_eulerr_data.csv";
writetable(T, setsfile);

Rpath = [FindRpath, filesep, 'Rscript.exe', '" "', '--vanilla '];
scriptfile = "C:\Users\fletcherpa\Documents\GitHub\scToolbox\eulerr.R";
pdf_file=fullfile(pwd,pdf_file);
command = char(strjoin([scriptfile,setsfile,options.shape,options.alpha,...
    options.width,options.height,options.fsz_counts,options.fsz_labels,options.rng_seed,pdf_file]," "));

commandline=['"', Rpath, command '"'];

[status, cmdout]=system(commandline);
if status~=0
    error(cmdout)
end
if options.verbose
    disp(cmdout)
end